### PR TITLE
Remove commonLabels from calculation of label-selectors

### LINKED ISSUE #1148: 🫐 🐛 Operator: Updating commonLabels does not seem to work
### What happened?

I set `commonLabels` in my CR. And, this is what I got: 

```console
Events:
  Type     Reason         Age   From                   Message
  ----     ------         ----  ----                   -------
  Warning  UpgradeFailed  8s    HelmReleaseReconciler  Helm upgrade failed for release redpanda/redpanda-broker with chart redpanda@5.7.37: cannot patch "redpanda-broker" with kind StatefulSet: StatefulSet.apps "redpanda-broker" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'ordinals', 'template', 'updateStrategy', 'persistentVolumeClaimRetentionPolicy' and 'minReadySeconds' are forbidden

Last Helm logs:

2024-04-04T14:49:59.015708Z: Patch Certificate "redpanda-broker-external-cert" in namespace redpanda
2024-04-04T14:49:59.040968736Z: Patch Certificate "redpanda-broker-letsencrypt-cert" in namespace redpanda
2024-04-04T14:49:59.065290115Z: Patch Certificate "redpanda-broker-selfsigned-cert" in namespace redpanda
2024-04-04T14:49:59.089878374Z: Patch Issuer "redpanda-broker-default-selfsigned-issuer" in namespace redpanda
2024-04-04T14:49:59.114574913Z: Patch Issuer "redpanda-broker-default-root-issuer" in namespace redpanda
2024-04-04T14:49:59.138430334Z: Patch Issuer "redpanda-broker-external-selfsigned-issuer" in namespace redpanda
2024-04-04T14:49:59.162118997Z: Patch Issuer "redpanda-broker-external-root-issuer" in namespace redpanda
2024-04-04T14:49:59.186072298Z: Patch Issuer "redpanda-broker-letsencrypt-root-issuer" in namespace redpanda
2024-04-04T14:49:59.208640725Z: Patch Issuer "redpanda-broker-selfsigned-root-issuer" in namespace redpanda
2024-04-04T14:49:59.36309197Z: warning: Upgrade "redpanda-broker" failed: cannot patch "redpanda-broker" with kind StatefulSet: StatefulSet.apps "redpanda-broker" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'ordinals', 'template', 'updateStrategy', 'persistentVolumeClaimRetentionPolicy' and 'minReadySeconds' are forbidden
  Normal  RollbackSucceeded  6s  HelmReleaseReconciler  Helm rollback to previous release redpanda/redpanda-broker.v1 with chart redpanda@5.7.37 succeeded
```

The release got roll back by Flux as a result. 

### What did you expect to happen?

See the labels apply to my statefulset pods. 

### How can we reproduce it (as minimally and precisely as possible)?. Please include values file.

by deploying a brand new cluster without setting `commonLabels`, then updating the CR to set it and re-deploying the cluster. 

### Anything else we need to know?

_No response_

### Which are the affected charts?

Redpanda, Operator

### Chart Version(s)

```console
❯ helm -n redpanda list
NAME             	NAMESPACE	REVISION	UPDATED                               	STATUS  	CHART          	APP VERSION
redpanda         	redpanda 	10      	2024-04-04 10:49:54.999923 -0400 EDT  	deployed	redpanda-0.1.1 	0.1.0
redpanda-broker  	redpanda 	5       	2024-04-04 14:50:18.75075024 +0000 UTC	deployed	redpanda-5.7.37	v23.3.10
redpanda-operator	redpanda 	1       	2024-04-01 18:14:29.732128 -0400 EDT  	deployed	operator-0.4.20	v2.1.15-23.3.7

```

### Cloud provider

<details>
Azure / AKS
</details>

### PR DESCRIPTION
**Pull request by alejandroEsc**
An additionalSelectorLabels map created, so that a custom set of label selectors can be created. 

fixes #1148

### CODE REVIEW

File: charts/redpanda/temp-template.yaml [hunk 1]
```diff
@@ -0,0 +1,1528 @@
+---
```
Comment (RafalKorepta) [inline charts/redpanda/temp-template.yaml:1]: Should this file be committed?

File: charts/redpanda/helpers.go [hunk 2]
```diff
@@ -92,7 +92,7 @@ func StatefulSetPodLabelsSelector(dot *helmette.Dot, statefulSet map[string]any)
 		"app.kubernetes.io/name":      Name(dot),
 	}
 
-	return helmette.Merge(commonLabels, defaults)
+	return helmette.Merge(AdditionalSelectorLabels, defaults)
```
Comment (RafalKorepta) [inline charts/redpanda/helpers.go]:
```suggestion
	return helmette.Merge(AdditionalSelectorLabels, values.CommonLabels, defaults)
```

That change could be backwards compatible.

File: charts/redpanda/Chart.yaml [hunk 3]
```diff
@@ -23,7 +23,7 @@ type: application
 # The chart version and the app version are not the same and will not track
 # together. The chart version is a semver representation of changes to this
 # chart.
-version: 5.7.41
+version: 5.7.42
```
Comment (RafalKorepta) [inline charts/redpanda/Chart.yaml:26]: If the first commit replaces commonLabels with new value, then I would suggest the following bump
```suggestion
version: 5.8.1
```

File: charts/kminion/README.md [hunk 4]
```diff
@@ -10,9 +10,6 @@ description: The most popular Open Source Kafka JMX to Prometheus tool by the cr
 
 This page describes the official Redpanda KMinion Helm Chart. In particular, this page describes the contents of the chart’s [`values.yaml` file](https://github.com/redpanda-data/helm-charts/blob/main/charts/kminion/values.yaml). Each of the settings is listed and described on this page, along with any default values.
 
-----------------------------------------------
-Autogenerated from chart metadata using [helm-docs v1.11.2](https://github.com/norwoodj/helm-docs/releases/v1.11.2)
-
```
Comment (RafalKorepta) [inline charts/kminion/README.md:15]: I think this shouldn't be removed.

File: charts/redpanda/values.yaml [hunk 5]
```diff
@@ -641,6 +641,7 @@ statefulset:
   #	any dedicated annotation.
   annotations: {}
   # -- Additional labels to be added to statefulset label selector.
+  # Caution should be taken that labels do exist on pods, see common labels.
```
Comment (RafalKorepta) [inline charts/redpanda/values.yaml:644]: I don't understand this comment.

File: charts/redpanda/helpers.go [hunk 6]
```diff
@@ -92,7 +92,26 @@ func StatefulSetPodLabelsSelector(dot *helmette.Dot, statefulSet map[string]any)
...
 	}
 
-	return helmette.Merge(commonLabels, defaults)
+	result := helmette.Merge(AdditionalSelectorLabels, defaults)
+	podLabels := StatefulSetPodLabels(dot, statefulSet)
+	if !IsMapSubset(podLabels, result) {
+		panic("error: computed label selector is not a subset of pod labels!")
+	}
+
+	return result
```
Comment (RafalKorepta) [inline charts/redpanda/helpers.go:101]: Why this is required?

File: charts/redpanda/helpers.go [hunk 7]
```diff
@@ -92,7 +92,26 @@ func StatefulSetPodLabelsSelector(dot *helmette.Dot, statefulSet map[string]any)
...
+}
+
+// IsMapSubset with the help of generics, determines if one map is a subset of another.
```
Comment (chrisseto) [inline charts/redpanda/helpers.go:104]: super nit: There's no need to mention generics in the comments as the function signature makes that obvious.

File: charts/redpanda/helpers.go [hunk 8]
```diff
@@ -92,7 +92,26 @@ func StatefulSetPodLabelsSelector(dot *helmette.Dot, statefulSet map[string]any)
 		"app.kubernetes.io/name":      Name(dot),
 	}
 
-	return helmette.Merge(commonLabels, defaults)
+	result := helmette.Merge(AdditionalSelectorLabels, defaults)
```
Comment (chrisseto) [inline charts/redpanda/helpers.go:95]: nit: could this variable be called `selectors` or something a bit more descriptive? The `IsMapSubset` line is a lot more difficult to understand as I had to keep referring to the definition of `result`.

File: charts/redpanda/temp-template.yaml (hunk 1 above)
Comment (chrisseto) [reply charts/redpanda/temp-template.yaml:1]: Looks like an accidental commit. Please remove.

File: charts/redpanda/helpers.go (hunk 2 above)
Comment (alejandroEsc) [reply charts/redpanda/helpers.go]: If we are doing a minor version bump, then let's not.

Comment (RafalKorepta) [reply charts/redpanda/helpers.go]: Agree with you

File: charts/redpanda/values_partial_test.go [hunk 9]
```diff
@@ -53,6 +54,11 @@ func TestPartialValuesRoundTrip(t *testing.T) {
 			unstructuredJSON, err := json.Marshal(unstructuredValues)
 			require.NoError(t, err)
 
+			t.Logf("structured: %s", string(structuredJSON))
+			fmt.Printf("structured: %s\n", string(structuredJSON))
```
Comment (chrisseto) [inline charts/redpanda/values_partial_test.go:58]: Left over `fmt.Printf`s? The `t.Log`s should be sufficient, no?

File: charts/redpanda/values.go [hunk 10]
```diff
@@ -232,9 +232,10 @@ type PodTemplate struct {
 }
 
 type Statefulset struct {
-	NodeAffinity   map[string]any `json:"nodeAffinity"`
-	Replicas       int            `json:"replicas" jsonschema:"required"`
-	UpdateStrategy struct {
+	AdditionalSelectorLabels map[string]string `json:"additionalSelectorLabels"`
```
Comment (chrisseto) [inline charts/redpanda/values.go:235]: Sorry, I don't recall if we talked about this already but is there any reason to _not` make this required (`jsonschema:"required"`)? It's nice to not have to worry about nulls when possible.

If we specify a default within `values.yaml` we won't require end users to make any changes as the JSONSchema is checked _after_ our `values.yaml` and the users' values are merged.

File: charts/redpanda/helpers.go [hunk 11]
```diff
@@ -113,11 +113,11 @@ func StatefulSetPodLabels(dot *helmette.Dot, statefulSet map[string]any) map[str
 		statefulSetLabels = values.Statefulset.PodTemplate.Labels
 	}
 
-	defults := map[string]string{
+	defaults := map[string]string{
 		"redpanda.com/poddisruptionbudget": Fullname(dot),
 	}
 
-	return helmette.Merge(statefulSetLabels, StatefulSetPodLabelsSelector(dot, nil), defults)
+	return helmette.Merge(statefulSetLabels, StatefulSetPodLabelsSelector(dot, nil), defaults)
```
Comment (chrisseto) [inline charts/redpanda/helpers.go:120]: hmmmm. Could passing in `nil` here potentially cause issues? If this is an upgrade and the existing statefulset had differing labels (IE the first if clause of `StatefulSetPodLabelsSelector` is `true`) then we might be generating incorrect labels here.

File: charts/redpanda/values_partial_test.go (hunk 9 above)
Comment (alejandroEsc) [reply charts/redpanda/values_partial_test.go:58]: oh no, i meant to remove that!

File: charts/redpanda/values.go (hunk 10 above)
Comment (alejandroEsc) [reply charts/redpanda/values.go:235]: not sure, but yeah we can make it required.

**Pull request merged**